File: 1_1_showdown/internal/domain/showdown.go
package domain

import "fmt"

const MaxTurn = 13

type Showdown struct {
	turn    int
	deck    *Deck
	players []Player
	table   map[PlayerName]Card // key is a HumanPlayer name
}

func NewShowdown(deck *Deck) Showdown {
	return Showdown{
		turn:    0,
		deck:    deck,
		players: []Player{},
		table:   make(map[PlayerName]Card, 4),
	}
}

func (s *Showdown) AddPlayer(p Player) {
	s.players = append(s.players, p)
}

func (s *Showdown) Start() {
	for _, p := range s.players {
		p.Naming()
	}

	s.deck.Shuffle()

	for s.deck.Size() > 0 {
		for _, p := range s.players {
			p.DrawCard(s.deck)
		}
	}

	// 這裡我更改遊戲規則，避免到時候手牌全部都是空的
	for s.turn < MaxTurn {
		// 第一個動作：決定要不要換牌
		for _, p := range s.players {
			p.TakeTurn1(p)
		}

		// 第二個動作：出牌
		for _, p := range s.players {
			p.TakeTurn2(p)
		}

		// find which player is bigger
		var biggerOne PlayerName
		for name, card := range s.table {
			fmt.Printf("Player: %s shows %s  \n", name, card.String())

			if biggerOne == "" {
				biggerOne = name
				continue
			}

			if card.Bigger(s.table[biggerOne]) {
				biggerOne = name
			}
		}

		// give him/her a nice point
		for _, p := range s.players {
			if p.GetName() == biggerOne {
				fmt.Printf("%s got a point!\n", biggerOne)
				p.GainPoint()
			}
		}

		s.turn++
	}

	// find who is the winner
	point, winner := 0, PlayerName("")
	for _, p := range s.players {
		fmt.Printf("Player gets %s point %d!\n", p.GetName(), p.GetPoint())
		if p.GetPoint() > point {
			point = p.GetPoint()
			winner = p.GetName()
		}
	}

	fmt.Printf("Winner is %s! Point is %d\n", winner, point)
}

func (s *Showdown) PutCardOnTable(c Card, p PlayerName) {
	s.table[p] = c
}

File: 1_1_cpp/domain/showdown.cpp

#include "showdown.h"

const int MaxTurn = 13;

using PlayerName = std::string;

Showdown::Showdown(Deck* d) {
    deck = d;
}

void Showdown::AddPlayer(IPlayer* p) {
    players.push_back(p);
}

void Showdown::Start() {
    for (IPlayer* p : players) {
        p->Naming();
    }

    deck->Shuffle();

    while (deck->Size() > 0) {
        for (IPlayer* p : players) {
            p->DrawCard(deck);
        }
    }

    // Modify the game rule to prevent all hands from being empty
    while (turn < MaxTurn) {
        // First action: decide whether to exchange hands
        for (IPlayer* p : players) {
            p->TakeTurn1();
        }

        // Second action: play a card
        for (IPlayer* p : players) {
            p->TakeTurn2();
        }

        // Find which player has the bigger card
        PlayerName biggerOne = "";
        for (const auto& pair : table) {
            Card c = pair.second;
            std::cout << "Player: " << pair.first << " shows " << c.String() << std::endl;

            if (biggerOne == "") {
                biggerOne = pair.first;
                continue;
            }

            if (c.Bigger(table[biggerOne])) {
                biggerOne = pair.first;
            }
        }

        // Give the winner a point
        for (IPlayer* p : players) {
            if (p->GetName() == biggerOne) {
                std::cout << p->GetName() << " got a point!" << std::endl;
                p->GainPoint();
            }
        }

        turn++;
    }

    // Find the winner
    int point = 0;
    PlayerName winner = "";
    for (IPlayer* p : players) {
        std::cout << "Player gets " << p->GetName() << " point " << p->GetPoint() << "!" << std::endl;
        if (p->GetPoint() > point) {
            point = p->GetPoint();
            winner = p->GetName();
        }
    }

    std::cout << "Winner is " << winner << "! Point is " << point << std::endl;
}
void Showdown::PutCardOnTable(const Card& c, const PlayerName& p) {
    table[p] = c;
}